src/chain.h

```diff
@@ -355,6 +355,19 @@ class CBlockIndex
     //! Efficiently find an ancestor of this block.
     CBlockIndex* GetAncestor(int height);
     const CBlockIndex* GetAncestor(int height) const;
+
+protected:
+    //! CBlockIndex should not allow public copy construction because equality
+    //! comparison via pointer is very common throughout the codebase, making
+    //! use of copy a footgun. Also, use of copies do not have the benefit
+    //! of simplifying lifetime considerations due to attributes like pprev and
+    //! pskip, which are at risk of becoming dangling pointers in a copied
+    //! instance.
+    //!
+    //! We declare these protected instead of simply deleting them so that
+    //! CDiskBlockIndex can reuse copy construction.
+    CBlockIndex(const CBlockIndex&) = default;
+    CBlockIndex& operator=(const CBlockIndex&) = default;
```

Comment:
Consider using the rule of five https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all.

Not sure (haven't tested), would it be better to define these separately in CBlockIndex (= delete) and CDiskBlockIndex (= default)?

Reply:
> Rule of 5

Sounds good, fixed.

> Not sure (haven't tested), would it be better to define these separately in CBlockIndex (= delete) and CDiskBlockIndex (= default)?

This doesn't work because to get a default copy constructor, the signature needs to look like `T(const T&) = default;`; because the CDiskBlockIndex constructor in question has a reference to a CBlockIndex pointer, we can't make use of a default copy constructor for CDiskBlockIndex. If we define one, we have to spell out all members from CBlockIndex, which is brittle.

Reply:
> operator= isn't needed as far as I can see

Good point, done.